### PR TITLE
Eigen::Ref support

### PR DESCRIPTION
**Pull request by fdarricau**
Partial support, some work remains to be done SolverState/OptimizationLogger to make it work, and then this commit (99fc832cb3dfd34b3624449eb944bdba01a55c6d) will have to be reverted.

### CODE REVIEW

File: include/roboptim/core/solver-state.hh [hunk 1]
```diff
@@ -82,9 +82,9 @@ namespace roboptim
     typedef P problem_t;
 
     /// \brief Import argument types from problem
-    typedef typename P::function_t::argument_t argument_t;
-    typedef typename P::function_t::argument_ref argument_ref;
-    typedef typename P::function_t::const_argument_ref const_argument_ref;
+    typedef typename P::vector_t argument_t;
+    typedef typename P::vector_t& argument_ref;
+    typedef const typename P::vector_t& const_argument_ref;
```
Comment (bchretien) [inline include/roboptim/core/solver-state.hh]: Why was this needed? The function [holds the proper typedefs](https://github.com/roboptim/roboptim-core/blob/master/include/roboptim/core/function.hh#L173).

**Comment by fdarricau** [comment]
The branch is much cleaner right now.

**Comment by bchretien** [comment]
I think the best course of action is to make a `3.0` branch and put these commits on it. Then we enable this branch on Travis, and do the same for other packages (cf. #84), while making sure that Travis uses the `3.0` branch for the dependencies (cf. [this](https://github.com/jrl-umi3218/jrl-travis/blob/master/before_install#L71)). Once we've got the green light from Travis for all the packages, we merge all the `3.0` branch to `master`.

**Comment by bchretien** [comment]
I'm closing this, the commits are currently on `3.0`.
